File: tree/binary-tree-vertical-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public class TreeNodeColumn {
        TreeNode node;
        int column;
        TreeNodeColumn(TreeNode node, int column) {
            this.node = node;
            this.column = column;
        }
    }
    
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result; 
        }
        
        Map<Integer, List<Integer>> columns = new HashMap<>();
        int min = 0;
        int max = 0;
        
        Deque<TreeNodeColumn> queue = new ArrayDeque<>();
        queue.addFirst(new TreeNodeColumn(root, 0));
        while (!queue.isEmpty()) {

            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNodeColumn treeNodeColumn = queue.pollFirst();
                TreeNode node = treeNodeColumn.node;
                int column = treeNodeColumn.column;
                
                min = Math.min(column, min);
                max = Math.max(column, max);
                
                if (!columns.containsKey(column)) {
                    columns.put(column, new ArrayList<>());
                }
                columns.get(column).add(node.val);
                
                if (node.left != null) {
                    queue.addLast(new TreeNodeColumn(node.left, column - 1));
                }
                
                if (node.right != null) {
                    queue.addLast(new TreeNodeColumn(node.right, column + 1));
                }
            }
        }
        
        for (int i = min; i <= max; i++) {
            result.add(columns.get(i));
        }
        
        return result;
    }
    
}

File: tree/binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root == None:
            return []
        map = {}
        self.traverse(root, map)
        res = []
        i = min(map.keys())
        while i in map.keys():
            res.append(map[i])
            i += 1
        return res

    def traverse(self, root: Optional[TreeNode], map: dict[int, list[int]]):
        if root == None:
            return
        cur = [(root, 0)]
        next = []
        while len(cur) > 0:
            for (node, col) in cur:
                if col not in map.keys():
                    map[col] = []
                map[col].append(node.val)
                if node.left != None:
                    next.append((node.left, col-1))
                if node.right != None:
                    next.append((node.right, col+1))
            cur = next
            next = []